### ISSUE TITLE
Mercurial projects not found

### ISSUE THREAD

**Issue by AlbertoFabbri93** [alefragnani/vscode-project-manager#438]
Issue Type: <b>Bug</b>

1. Set Project Manager › Hg: Base Folders
2. Enable Project Manager: Group List

Expected result:
In the sidebar a new group with the projects that use mercurial is displayed

Actual result:
Nothing change, only favourites and git projects are displayed

Extension version: 11.3.1
VS Code version: Code 1.51.1 (e5a624b788d92b8d34d1392e4c4d9789406efe8f, 2020-11-10T23:34:32.027Z)
OS version: Windows_NT x64 10.0.19042

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i5-8250U CPU @ 1.60GHz (8 x 1800)|
|GPU Status|2d_canvas: enabled<br>flash_3d: enabled<br>flash_stage3d: enabled<br>flash_stage3d_baseline: enabled<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>oop_rasterization: disabled_off<br>opengl: enabled_on<br>protected_video_decode: enabled<br>rasterization: enabled<br>skia_renderer: disabled_off_ok<br>video_decode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|7.84GB (1.06GB free)|
|Process Argv|--crash-reporter-id 7cd19320-ee8a-45d8-b96c-f694d3262d4d|
|Screen Reader|no|
|VM|0%|
</details>

**Comment by alefragnani** [comment]
Hi @AlbertoFabbri93 ,

You don’t need to set “group list”, in order to display a new group/panel/view as you see for Favorites and Git. This should be automatically displayed if you set the `baseFolders` and at least one repo is found. The “groupList` only affects the Command Palete. Btw, the “detector” simply checks for the existence of `.hg` or `.hgrc ` folder, nothing else.

Did you see a `Refresh Projects ...` message after you updated the settings? If not, try running the `Project Manager: Refresh Projects` command, to force the auto-detection .

Hope this helps

**Comment by AlbertoFabbri93** [comment]
Hi @alefragnani,

Thank you for the quick reply.
`group list` is not necessary, you are right, but I don't yet understand what it is used for.
I checked and both my mercurial projects have an `.hg` folder.
I tried to refresh the projects list with the commands: `Refresh Projects`, `Refresh Any Projects` and `Refresh Mercurial Projects` but nothing changed. What's the difference between the first two?

Any other advice would be very appreciated.

Thanks!

**Comment by alefragnani** [comment]
Hi @AlbertoFabbri93 

The `Group` setting is used to display the projects _grouped by its kind_ in the `Command Palette`. It's not a _treeview-like_ approach, but just a _all favorites_, then _all Git_, then _all SVN_ and so on..

The `Refresh Projects` should refresh _all detectable projects_, so in the end, it calls the `Refresh Mercurial` too.

Could you show your settings, and your folder, where you have the Mercurial repos? Right now the extension still doesn't have a _log_, so there is nowhere to look for _why it is not working_. I have a almost ready solution for that, but still no ETA.

Hope this helps

**Comment by AlbertoFabbri93** [comment]
Thank you for the explanations!
This is my `settings.json `file:
```
{
    // EDITOR
    "editor.copyWithSyntaxHighlighting": false,
    "editor.fontFamily": "Fira Code",
    "editor.fontLigatures": true,
    "editor.suggestSelection": "first",

    // EXPLORER
    "explorer.confirmDelete": false,

    // TERMINAL
    "terminal.external.windowsExec": "C:\\Users\\albif\\AppData\\Local\\Microsoft\\WindowsApps\\wt.exe",
    "terminal.integrated.automationShell.windows": null,
    "terminal.integrated.cursorBlinking": true,
    "terminal.integrated.shell.windows": "C:\\Program Files\\PowerShell\\7\\pwsh.exe",

    // WORKBENCH
    "workbench.startupEditor": "welcomePage",

    // GIT
    "git.autofetch": true,
    "git.confirmSync": false,

    // PROJECT MANAGER
    "projectManager.groupList": true,
    "projectManager.git.baseFolders": [
        "C:\\Users\\albif\\Documents"
    ],
    "projectManager.hg.baseFolders": [
        "C:\\Users\\albif\\Documents"
    ],
    
    // FILE HEADER
    "psi-header.config": {
        "forceToTop": true,
        "blankLinesAfter": 1,
        "author": "Alberto Fabbri"
    },
    "psi-header.lang-config": [
        {
            "language": "python",
            "begin": "",
            "end": "",
            "prefix": "# "
        },
        {
            "language": "java",
            "prefix": "",
        },
    ],
    "psi-header.templates": [
        {
            "language": "*",
            "template": [
                "    _    _ _     _",
                "   / \\  | | |__ (_)",
                "  / _ \\ | | '_ \\| |",
                " / ___ \\| | |_) | |",
                "/_/   \\_\\_|_.__/|_|",
                "Author: <<author>>",
                "GitLab: https://gitlab.com/Albi-F/",
                "GitHub: https://github.com/AlbertoFabbri93"
            ]
        },
    ],

    // PYTHON
    "AREPL.delay": 500,
    "autoDocstring.generateDocstringOnEnter": false,
    "autoDocstring.startOnNewLine": true,
    "python.languageServer": "Pylance",
    "python.testing.pytestEnabled": true,

    // JAVA
    "java.configuration.checkProjectSettingsExclusions": false,

    // LATEX
    "latex-workshop.view.pdf.viewer": "tab",

    // OTHERS
    "vsintellicode.modify.editor.suggestSelection": "automaticallyOverrodeDefaultValue",
}
```
I created a minimal mercuarial repo structured this way:

```
merctest
└───.hg
    ├───cache
    ├───store
    └───wcache
```

The merctest folder is inside Documents.

**Comment by alefragnani** [comment]
I think I found the reason why your repos are not detected.

The extension doesn't do any Mercurial API/CLI check, while detecting Mercurials repo. Instead, it just look for folders which contains a `.hg` subfolder, with a `hgrc` file inside (`.hg/hgrc`). In your case, I don't see a `hgrc` file, so it is not recognized as Mercurial repo.

I'm not a Mercurial expert. In fact, I have little use of Mercurial. I have a couple of repos, cloned from BitBucket years ago, but never created one locally. I found some references to `hgrc` file being the config file, which contains things like the remote url of your repo. Being local _yet_, may be the reason why it is not present.

I didn't find any documentation/resources that gives me solid documentation on how a _minimal_ Mercurial repo should look like, but based on you sample, it appears just `.hg` would do the trick. Do you have any docs about it?

Out of curiosity, could you say the contents of `cache`, `store` and `wcache` files?

**Comment by AlbertoFabbri93** [comment]
I'm not used to Mercurial either so I have no idea what a minimal repo should look like but on my computer I have the Mercurial repo of Mozilla Firefox and it too doesn't containt the `hgrc` file in the `.hg` folder. I tried to add it to the `merctest` folder and now it shows up in VS Code.

I'm sorry, the tree I posted is misleading, `cache`, `store `and `cache `are empty folders. Hovewer the .hg folder contains two files:

**`00changelog.i`**:
``` dummy changelog to prevent using the old repo layout```

**`requires`**:
```dotencode
fncache
generaldelta
revlogv1
sparserevlog
store```
